### ISSUE TITLE
Group use statements

### ISSUE THREAD

**Issue by thekid** [xp-framework/compiler#108]
Currently, all `use` statements are emitted standalone:

```php
// Source code
use lang\{Value, Runnable};

// Result
new UseStatement(null, ['lang\\Value' => null, 'lang\\Runnable' => null], $line);

// Emitted as
use lang\Value; use lang\Runnable;
```

The emitted code could be more efficient when using [grouped use syntax](https://wiki.php.net/rfc/group_use_declarations) available since PHP 7.0.

**Comment by thekid** [comment]
One option could be to parse grouped use statements into a new node `GroupedUseStatement`. This would however be a BC break.

**Comment by thekid** [comment]
Another idea would be to declare a `base` member with the base namespace (*everything before the opening curly brace, `lang` in the above example*) and set this member. Emitters knowing about this member could then fairly easily emit the grouped syntax.

**Comment by thekid** [comment]
> One option could be to parse grouped use statements into a new node `GroupedUseStatement`. This would however be a BC break.

This might be the cleaner alternative though, we might keep this here for the next major release in this case.

**Comment by Danon** [comment]
@thekid Hello, are you looking for contributors?

**Comment by thekid** [comment]
> @thekid Hello, are you looking for contributors?

Sure, help is always welcome 👍 

Did you have a specific idea in mind what you would like to contribute or are you be looking for things to be done? I have a small amount of issues tagged [help wanted](https://github.com/xp-framework/compiler/labels/help%20wanted),  and I'm keen on receiving feedback as to what you would want to see in this piece of software!

**Comment by Danon** [comment]
> > @thekid Hello, are you looking for contributors?
> 
> Sure, help is always welcome 👍
> 
> Did you have a specific idea in mind what you would like to contribute or are you looking for things to be done? I have a small amount of issues tagged [help wanted](https://github.com/xp-framework/compiler/labels/help%20wanted), and I'm keen on receiving feedback as to what you would want to see in this piece of software!

Yes and no.

First of all, I'm a library maintainer myself, and keeping backwards comaptible changes is hard. So when I saw your project, I thought to myself: "Hey, if this project kicks, I can develop my library using newest PHP 8.1 features, and still maintain full PHP 7 support.". So I saw areas where I could use your project.

Second thing is, my biggest advantage is being detail-oriented and finding corner-cases, I think that might be useful in this project.

And third, my biggest skill is writing good unit tests, and I think you might Look kindly on it.

**Comment by thekid** [comment]
> I can develop my library using newest PHP 8.1 features, and still maintain full PHP 7 support

Yes, that would be possible! Maybe a word of warning: The generated code is quite ugly, does not contain any comments and contains some XP Framework-related meta data:

```bash
$ echo '<?php class T { }' | xp compile -
<?php  class T{static function __init() {\xp::$meta['T']= /* ... */ }} \T::__init();;
```

One idea to make this more generically usable would be to make rewriting api doc comments and annotations to meta data optional, e.g. by regarding it an optimization (*XP Framework will parse it from the class' sourcecode if not present*). This way, your library wouldn't need an additional runtime dependency.

Maybe this would be a good starting point for a joint effort?

> Second thing is, my biggest advantage is being detail-oriented and finding corner-cases, I think that might be useful in this project.

While working on the above project together, we would certainly find a lot of these 😉

> And third, my biggest skill is writing good unit tests, and I think you might Look kindly on it.

Here's some statistics on coverage - the classes used by the command line tool could need some coverage love:

![image](https://user-images.githubusercontent.com/696742/124460531-c334a080-dd8f-11eb-935b-7dfa52250d71.png)

**Comment by thekid** [comment]
> the classes used by the command line tool could need some coverage love

I've taken care of this on the meantime 😉 - but help on making the compiled code independent of XP core is greatly appreciated, see issue #116

**Comment by Danon** [comment]
@thekid So what's the idea behind this project?

Did you write some ideas/concepts somewhere, so I can get familiar? 

For example, if there are typed properties in PHP 8.0, but not in PHP 7.0, will you just ignore the types? Or will you write some polyfill that would validate the type?

**Comment by thekid** [comment]
> So what's the idea behind this project?

I've written up a document on this @Danon - see here: https://github.com/xp-framework/compiler/wiki/Compiler-design. 

> Or will you write some polyfill that would validate the type?

I'm open to simulate typed properties with a combination of __get and __set - I think the performance overhead is okay-ish. Want to give this a try in form of a pull request?

**Comment by thekid** [comment]
> The emitted code could be more efficient when using [grouped use syntax](https://wiki.php.net/rfc/group_use_declarations) available since PHP 7.0.

However, the **emitting** code is much more complex if we try to do this without BC breaks.
